### PR TITLE
Save drawing

### PR DESCRIPTION
**Pull request by R0ss**
Still working on having the new graphs be updated with a title that include the graph's Id number...

### CODE REVIEW

File: dependencies/HaXml-1.25.3/src/Text/XML/HaXml/Wrappers.hs [hunk 1]
```diff
@@ -31,7 +31,7 @@ fix2Args :: IO (String,String)
 fix2Args = do
   args <- getArgs
   when ("--version" `elem` args) $ do
-      putStrLn $ "part of HaXml-"++ VERSION
+      putStrLn $ "part of HaXml-"
```
Comment (Ian-Stewart-Binks) [inline dependencies/HaXml-1.25.3/src/Text/XML/HaXml/Wrappers.hs]: Why did you modify HaXml?

File: hs/Database/Tables.hs [hunk 2]
```diff
@@ -140,6 +141,13 @@ FacebookTest
 
 -- ** TODO: Remove these extra types and class instances
 
+-- | JSON SVG data
+data SvgJSON =
+    SvgJSON { texts :: [Text],
```
Comment (Ian-Stewart-Binks) [inline hs/Database/Tables.hs]: See `TODO` above. Is it possible to modify the database schema instead of creating the extra type?

File: public/js/draw/draw.js [hunk 3]
```diff
@@ -389,3 +390,98 @@ function regionClicked(e) {
...
+    var shapes = [];
+    var paths = [];
+    var svgElements = document.getElementById('mySVG').children;
+    for (var i = 0; i < svgElements.length; i++) {
+        if (svgElements[i].tagName == "g" && svgElements[i].id != "regions") {
+            var textObj = "";
+            var gText = svgElements[i].querySelector("text");
+            if (gText !== null) {
+                textObj = {
+                    "graph"     : parseInt(gId),
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: `'`

File: public/js/draw/draw.js [hunk 4]
```diff
@@ -389,3 +390,98 @@ function regionClicked(e) {
...
+                    "target"    : getClosestText(pathCoords[1], shapes)
+            });
+        }
+    }
+    return JSON.stringify({texts:texts, shapes:shapes, paths:paths});
+}
+
+
+/**
+ * getPathCoords(pathCoords)
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: Putting the function name in the JSdoc is not only redundant, but also bad from a maintainability and consistency standpoint. When the function is renamed in the future, the JSdoc might not be changed.

File: hs/Database/CourseInsertion.hs [hunk 5]
```diff
@@ -11,16 +11,40 @@ into the database. These functions are used as helpers for the WebParsing module
...
+-- | Insert SVG graph and return its id
+insertGraph :: IO Response
+insertGraph = do
+       gId <- runSqlite databasePath $ insert $ Graph "User generated graph" 256 256
+       return $ createJSONResponse gId
+
+-- | Inserts SVG graph data into Texts, Shapes, and Paths tables
+saveGraphJSON :: String -> IO Response
+saveGraphJSON jsonStr = do
+  print jsonStr
```
Comment (Ian-Stewart-Binks) [inline hs/Database/CourseInsertion.hs]: Why is it printed?

File: dependencies/HaXml-1.25.3/src/Text/XML/HaXml/Wrappers.hs (hunk 1 above)
Comment (R0ss) [inline dependencies/HaXml-1.25.3/src/Text/XML/HaXml/Wrappers.hs]: These changes allow `stack repl` to work, because `stack` had a problem with `VERSION`

File: hs/Database/CourseInsertion.hs (hunk 5 above)
Comment (R0ss) [inline hs/Database/CourseInsertion.hs]: Argg, it was debugging code...

File: public/js/draw/draw.js (hunk 4 above)
Comment (R0ss) [inline public/js/draw/draw.js]: Good point. I'll fix that.

File: hs/Database/Tables.hs (hunk 2 above)
Comment (R0ss) [inline hs/Database/Tables.hs]: It might be possible to delete it entirely. I only used it to use it with: `Just (SvgJSON texts shapes paths) -> do 
`

File: hs/Database/CourseInsertion.hs [hunk 6]
```diff
@@ -11,16 +11,40 @@ into the database. These functions are used as helpers for the WebParsing module
...
+import Util.Happstack (createJSONResponse)
+import Database.Persist.Sqlite (fromSqlKey, update, insertMany_, insert_, insert, SqlBackend, (=.), (==.), updateWhere, runSqlite)
 import Database.Tables
+import Data.Aeson
+
+-- | Insert SVG graph and return its id
+insertGraph :: IO Response
+insertGraph = runSqlite databasePath $ do
+    gId <- insert $ Graph "" 256 256
+    update gId [GraphTitle =. ((++) "User generated graph " $ show $ fromSqlKey gId)]
```
Comment (R0ss) [inline hs/Database/CourseInsertion.hs]: A new graph is inserted, and persistent generates a new `GraphId` for it, which is used to update the graph's title.

File: public/js/draw/draw.js (hunk 3 above)
Comment (R0ss) [inline public/js/draw/draw.js]: ?

Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: https://github.com/Courseography/courseography/pull/790#discussion_r53579355

Comment (R0ss) [inline public/js/draw/draw.js]: Ah, I see.

File: dependencies/HaXml-1.25.3/src/Text/XML/HaXml/Wrappers.hs (hunk 1 above)
Comment (david-yz-liu) [inline dependencies/HaXml-1.25.3/src/Text/XML/HaXml/Wrappers.hs]: @R0ss These changes don't belong in this pull request; if you want, you can submit them separately.

File: hs/Response/Post.hs [hunk 7]
```diff
@@ -115,18 +115,14 @@ checkPost =
...
                 H.div ! A.id "specextra" ! A.class_ "more-info" $ do
                     H.input ! A.type_ "text"
                     H.input ! A.type_ "text"
                     H.input ! A.type_ "text"
                     H.input ! A.type_ "text"
             H.div ! A.id "spec_misc" $ do
-                H.p ! A.class_ "code" $ "Any from this list: CSC301H, CSC318H, CSC404H, CSC411H, CSC418H, CSC420H, \
-                    \CSC428H, CSC454H, CSC485H, CSC490H, CSC491H, CSC494H, or PEY (0.5 FCEs) \
-                    \ ** Note: Type 'PEY' for Check my POSt to recognize it **"
+                H.p ! A.class_ "code" $ "Any from this list: CSC301H, CSC318H, CSC404H, CSC411H, CSC418H, CSC420H,  CSC428H, CSC454H, CSC485H, CSC490H, CSC491H, CSC494H, or PEY (0.5 FCEs)   ** Note: Type 'PEY' for Check my POSt to recognize it **"
```
Comment (david-yz-liu) [inline hs/Response/Post.hs]: Same comment as HaXml - please don't include these changes in this pull request.

File: public/js/draw/draw.js [hunk 8]
```diff
@@ -389,3 +390,95 @@ function regionClicked(e) {
...
+ */
+function convertSvgToJson(gId) {
+    'use strict';
+
+    var texts = [];
+    var shapes = [];
+    var paths = [];
+    var svgElements = document.getElementById('mySVG').children;
+    for (var i = 0; i < svgElements.length; i++) {
+        if (svgElements[i].tagName == 'g' && svgElements[i].id != 'regions') {
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: Use `===` and `!==`

File: public/js/draw/setup.js [hunk 9]
```diff
@@ -86,6 +86,30 @@ $('#finish-region').click(function () {
     finishRegion();
     });
 
+$('#save-graph').click(function () {
+        $.ajax({
```
Comment (david-yz-liu) [inline public/js/draw/setup.js]: Let's indent this by 4 spaces.

File: hs/Database/CourseInsertion.hs [hunk 10]
```diff
@@ -11,16 +11,46 @@ into the database. These functions are used as helpers for the WebParsing module
...
+insertGraph :: String -> IO Response
+insertGraph nameStr = runSqlite databasePath $ do
+    nameExists <- selectFirst [GraphTitle ==. nameStr] []
+    case nameExists of
+      Nothing -> do
+        gId <- insert $ Graph nameStr 256 256
+        return $ createJSONResponse gId
+      Just _ -> do
+        gId <- insert $ Graph "" 256 256
+        update gId [GraphTitle =. ((++) "User generated graph " $ show $ fromSqlKey gId)]
```
Comment (david-yz-liu) [inline hs/Database/CourseInsertion.hs]: For now, let's not insert anything in the case that the graph name is already taken.

File: hs/Database/Tables.hs (hunk 2 above)
Comment (david-yz-liu) [inline hs/Database/Tables.hs]: I think we need a separate type to create a `FromJSON` instance.

File: hs/DevelopmentConfig.hs [hunk 11]
```diff
@@ -123,4 +123,4 @@ holidays = ["20151012T", "20151109T", "20151110T",
 
 -- | Enable Facebook integration. Should only be true on the production server.
 enableFb :: Bool
-enableFb = False
+enableFb = True
```
Comment (david-yz-liu) [inline hs/DevelopmentConfig.hs]: Don't change this.

File: public/js/draw/setup.js [hunk 12]
```diff
@@ -86,19 +86,47 @@ $('#finish-region').click(function () {
     finishRegion();
     });
 
+$('#save-graph').click(function () {
```
Comment (david-yz-liu) [inline public/js/draw/setup.js]: There should just be one AJAX request here, not two.

File: public/js/draw/setup.js [hunk 13]
```diff
@@ -86,19 +86,47 @@ $('#finish-region').click(function () {
...
-            data: {gid : $('#area-of-study').val()},
-            dataType: 'json', 
-            success: function(data) {
-                $('#json-data').html('<pre>' + JSON.stringify(data) + '<pre>');
-            },
-            error: function(xhr, status, err) {
-                console.error('graphs', status, err.toString());
+$('#save-graph').click(function () {
+    $.ajax({
+        url: 'insert-graph',
```
Comment (david-yz-liu) [inline public/js/draw/setup.js]: You should be able to do this in one AJAX request, not two. You'll need to set graph IDs on the server, not here.

Comment (R0ss) [inline public/js/draw/setup.js]: The only reason I separate it into two Ajax requests is because of the situation where the user enters an invalid graph name multiple times. I don't want to send the us in data until a valid name is entered.

Comment (R0ss) [inline public/js/draw/setup.js]: The json data*

Comment (david-yz-liu) [inline public/js/draw/setup.js]: That's not a bad thought, but the way this is set up now, it is possible for a graph to be inserted while its corresponding elements are not. E.g., if the JSON data is not parsed properly, which actually also happened to me while I was testing this.

It will be possible in the future to check for duplicate names through some other method.

**Comment by david-yz-liu** [comment]
Also make sure to do a pull from Courseography's master branch so that you don't have any merge conflicts, and test your code thoroughly.

**Comment by R0ss** [comment]
Ah, I see.

File: public/js/draw/draw.js [hunk 14]
```diff
@@ -389,3 +390,107 @@ function regionClicked(e) {
...
+        return a[1] === b[1] ? 0 : a[1] < b[1] ? -1: 1
+    })[0][0];
+}
+
+
+/**
+ * saveGraph(jsonData)
+ * Insert JSON data into Persistent graph table.
+ */
+function saveGraph(jsonData) {
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: I think you can remove this function.

File: public/js/draw/draw.js [hunk 15]
```diff
@@ -389,3 +390,107 @@ function regionClicked(e) {
...
+                    'graph'     : parseInt(gId),
+                    'id_'       : '',
+                    'pos'       : [ parseFloat(gRect.getAttribute('x')),
+                                    parseFloat(gRect.getAttribute('y'))
+                                    ],
+                    'width'     : parseFloat(gRect.getAttribute('width')),
+                    'height'    : parseFloat(gRect.getAttribute('height')),
+                    'fill'      : 'none',
+                    'stroke'    : '',
+                    'text'      : [textObj],
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: Use your `texts` variable from above.

File: public/js/draw/draw.js [hunk 16]
```diff
@@ -389,3 +390,107 @@ function regionClicked(e) {
...
+    var texts = [];
+    var shapes = [];
+    var paths = [];
+    var svgElements = document.getElementById('mySVG').children;
+    for (var i = 0; i < svgElements.length; i++) {
+        if (svgElements[i].tagName === 'g' && svgElements[i].id !== 'regions') {
+            var textObj = '';
+            var gText = svgElements[i].querySelector('text');
+            if (gText !== null) {
+                textObj = {
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: You don't need a temporary variable at all - just push the object literal.

File: public/js/draw/draw.js [hunk 17]
```diff
@@ -389,3 +390,107 @@ function regionClicked(e) {
...
+        } else if (svgElements[i].tagName === 'path') {
+            var pathCoords = getPathCoords(svgElements[i].getAttribute('d'));
+            paths.push({
+                    'graph'     : parseInt(gId),
+                    'id_'       : '',
+                    'points'    : pathCoords,
+                    'fill'      : 'none',
+                    'stroke'    : '',
+                    'isRegion'  : false,
+                    'source'    : getClosestText(pathCoords[0], shapes),
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: It may be that `getClosestText` returns undefined, when the source/target node doesn't have a label.

This is a larger problem to deal with later, but for now please just set this function to return the empty string rather than undefined (this will fix another parsing problem).

File: app/Database/CourseInsertion.hs [hunk 18]
```diff
@@ -11,16 +11,35 @@ into the database. These functions are used as helpers for the WebParsing module
...
+import Data.Aeson
+
+-- | Inserts SVG graph data into Texts, Shapes, and Paths tables
+saveGraphJSON :: String -> String -> IO Response
+saveGraphJSON jsonStr nameStr = do
+    let jsonObj = decode $ BSL.pack jsonStr
+    case jsonObj of
+        Nothing -> return $ toResponse ("Error" :: String)
+        Just (SvgJSON texts shapes paths) -> do
+            response <- runSqlite databasePath $ do insertMany_ texts
```
Comment (david-yz-liu) [inline app/Database/CourseInsertion.hs]: style comment: start the first action after the `do` on a new line, so you only indent 4 spaces

File: app/Database/CourseInsertion.hs [hunk 19]
```diff
@@ -11,16 +11,35 @@ into the database. These functions are used as helpers for the WebParsing module
...
+                                                    insertMany_ shapes
+                                                    insertMany_ paths
+                                                    gId <- insert $ Graph nameStr 256 256
```
Comment (david-yz-liu) [inline app/Database/CourseInsertion.hs]: You're on the right track, but you should really use record updating.

If you have a text `t` and you want to change the graph field, keep in mind that the actual name of the field is `textGraph`, so you can do something like `t { textGraph = gId }`. Do that for all of `texts`, `shapes`, and `paths` before inserting into the database, so you won't have to later do an update.
